src/log.rs

```diff
@@ -0,0 +1,396 @@
+            }
+        }
+    }
+    let callback = if func.is_some() {
+        Some(func_func as _)
+    } else {
+        None
+    };
+    let func = func.map(|f| Box_::new(Box_::new(f)));
+    match PRINT_HANDLER.lock() {
+        Ok(mut handler) => {
+            *handler = unsafe { ::std::mem::transmute(func) };
+        }
+        Err(_) => {
+            // should we log something?
+        }
+    }
+    unsafe { glib_sys::g_set_printerr_handler(callback) };
+}
+
+static PRINTERR_HANDLER: Lazy<Mutex<Option<Box_<Box_<dyn Fn(GString) + Send + Sync + 'static>>>>> =
+    Lazy::new(|| Mutex::new(None));
+
+pub fn set_printerr_handler<P: Fn(GString) + Send + Sync + 'static>(func: Option<P>) {
+    unsafe extern "C" fn func_func(string: *const libc::c_char) {
+        match PRINTERR_HANDLER.lock() {
+            Ok(handler) => {
+                if let Some(ref handler) = *handler {
+                    let string: GString = from_glib_borrow(string);
+                    (*handler)(string)
+                } else {
+                    panic!("PRINTERR_HANDLER cannot be None!");
+                }
+            }
+            Err(_) => {
+                // should we log something here?
+            }
+        }
+    }
+    let callback = if func.is_some() {
+        Some(func_func as _)
+    } else {
+        None
+    };
+    let func = func.map(|f| Box_::new(Box_::new(f)));
+    match PRINTERR_HANDLER.lock() {
+        Ok(mut handler) => {
+            *handler = unsafe { ::std::mem::transmute(func) };
+        }
+        Err(_) => {
+            // should we log something?
+        }
+    }
+    unsafe { glib_sys::g_set_printerr_handler(callback) };
+}
+
+pub fn log_set_default_handler<P: Fn(&str, &LogLevelFlags, &str) + Send + Sync + 'static>(
+    log_func: P,
+) {
+    let log_func_data: Box_<P> = Box_::new(log_func);
+    unsafe extern "C" fn log_func_func<
+        P: Fn(&str, &LogLevelFlags, &str) + Send + Sync + 'static,
+    >(
+        log_domain: *const libc::c_char,
+        log_level: glib_sys::GLogLevelFlags,
+        message: *const libc::c_char,
+        user_data: glib_sys::gpointer,
+    ) {
+        let log_domain: GString = from_glib_borrow(log_domain);
+        let log_level = from_glib(log_level);
+        let message: GString = from_glib_borrow(message);
+        let callback: &P = &*(user_data as *mut _);
+        (*callback)(log_domain.as_str(), &log_level, message.as_str());
+    }
+    let log_func = Some(log_func_func::<P> as _);
+    let super_callback: Box_<P> = log_func_data;
+    unsafe {
+        glib_sys::g_log_set_default_handler(log_func, Box_::into_raw(super_callback) as *mut _);
+    }
+}
+
+/// Macro used to log using GLib logging system. Is uses [g_log].
+///
+/// [g_log]: https://developer.gnome.org/glib/stable/glib-Message-Logging.html#g-log)
+///
+/// Example:
+///
+/// ```no_run
+/// use glib::{LogLevelFlags, g_log};
+///
+/// g_log!(None, LogLevelFlags::FLAG_RECURSION, "test");
```

Comment:
In addition you probably want macros like `g_error!("domain", "foo {}", 1)` for each of the levels.